### ISSUE TITLE
Using user channel

### ISSUE THREAD

**Issue by johnhabbard** [danpaquin/gdax-python#226]
Hi gurus,

I am trying to use channel 'user'. I did all as described in the documentation, but I am not getting any user messages. I tried to run the WebsocketClient subscribed to channel 'user' and then add some trading requests at https://www.gdax.com/trade. Should I get a message on the channel then? I am not getting any. What am I doing wrong?

Could someone post an example of code that worked for them. I could not find any example for channel 'user' so far on the internet.

Basically, I wish to have a way to be notified when some of my trade request executed. If I use channel 'full', for example, it sends too many messages and some are out of sequence. So, this does not make me feel confident that I will get notified on my trade executed. This is why I tried channel 'user'.

Any ideas?
Thanks a lot in advance!

**Comment by lirenz** [comment]
1) Make sure you're authenticated
2) The user channel pushes data as the "full" channel (except it includes ONLY your updates)
(so matches, opens, etc). So type won't be "user"

**Comment by johnhabbard** [comment]
Yes, I've done all that, it is in the documentation. Did is ever work for you in real?

**Comment by lirenz** [comment]
yes, it works for me (I've tested it by posting a deep out order). This is on live (non-sandbox). You might want to pull the development version of the code though (ie, I'm using the straight github version, not the pip)

**Comment by lirenz** [comment]
Can you post the "on_open" part of your code?

**Comment by johnhabbard** [comment]
I am glad, Lawrence, that it works for you, then we should be able to figure it out. Thanks a lot for helping me. Here is my on_open:
```
	def on_open(self):
		global auth_key, auth_secret, auth_passphrase
		self.url = "wss://ws-feed.gdax.com/"
		self.products = ['BTC-USD'] # I have a variable here, which works fine for other channels
		self.channels = ['user'] # works fine for ['level2', 'full', 'ticker']
		self.auth = True
		self.api_key = auth_key
		self.api_secret = auth_secret
		self.api_passphrase = auth_passphrase
```
Same auth* vars are used for gdax.AuthenticatedClient and work fine. They were created with permissions: view, trade.

**Comment by johnhabbard** [comment]
I also get the first message of type 'subscriptions' showing {'name':'user', 'product_ids':['BTC-USD']}
However, after the first message it stays silent no matter what I try.

It times out after 60 seconds, because ping-part of the code does not get a chance to be executed, but this is what I can fixed. I run another separate thread which does ping self.ws.ping("keepalive") and then it does not time out. But I do not get any messages except first 'subscriptions' in any case.

**Comment by johnhabbard** [comment]
BTW, I do use the latest version from git: https://github.com/danpaquin/gdax-python/archive/master.zip 
I have already learned its code pretty well, but no luck on channel 'user' ...

**Comment by johnhabbard** [comment]
>I've tested it by posting a deep out order

Lawrence, did you put order using this python API in the same session, or did you use  https://www.gdax.com/trade to manually submit a test order? The latter is what I did. I have not tried sending it within python. It will probably work, but I want to be able to start monitoring at any time and get notified of an order executed, which was previously submitted either manually via https://www.gdax.com/trade or from a previous session of python API. Is this how you use it?

**Comment by lirenz** [comment]
I put in a trade manually. My on_open:

    def on_open(self):
        self.url = "wss://ws-feed.gdax.com/"

        self.products = ['BTC-USD']
        self.channels = ["level2", "ticker", "user"]

        self.auth = True
        self.api_key = config.gdax_key
        self.api_secret = config.gdax_secret
        self.api_passphrase = config.gdax_passphrase

     def on_message(self, message):
        product_id = message.get('product_id', 'na')
        message_type = message.get('type', 'na')

        if (message_type == 'received' or message_type == 'open' or message_type == 'done' or message_type == 'match' or message_type == 'change'):
            side = message.get('side', None)
            price = message.get('price', None)
            size = message.get('size', None)
            print 'here'

**Comment by johnhabbard** [comment]
Wow, I am totally puzzled: this still does not work for me. I get 'ticker' and 'l2update', but nothing else.
Would it be too much work for you to try to remove "level2" and "ticker" from your on_open and see if you still get 'user' messages? It should timeout after 60 seconds though. You would need to submit a test trade within 60 seconds after the start.

What permissions did you choose when created API key? My permissions are 'view' and  'trade'. I also tried an older key with view, trade, transfer, bypass-2fa, which works in the same way. I have not tried to enable permission 'manage', which does not seem relevant.

Permissions are the only difference between our codes now, which I can still think of.

**Comment by johnhabbard** [comment]
Here is something new. I tried to change one symbol in API key to make it invalid and used it in on_open. I did not get any error messages and all worked the same way. This makes me think that I am actually not authenticated properly.

Could you try the same and see if you get an error message? If yes, what does your message say?

Meanwhile, I'll look through the API code to see where I can put debugging output to make sure that I am authenticated properly.

**Comment by johnhabbard** [comment]
Wow. Success! I downloaded the development version from github and discovered that authentication code is quite different there from the code I had in my version. However, the development version did not work for me for some reason. Then I discovered that the latest master is also different from what I had and the new autherntication code from development version is already there. I tried then the latest master and it now works as expected ;)

Thank you very much, Lawrence, @lirenz 👍 , for your inspirational input. I would not know what to do without your notes. Much appreciated!
